File: chap1/sec1.5/prob1.10/sol_findPi.m
% Settings for Matlab
close all; clear;
set( 0, 'defaultAxesTickLabelInterpreter', 'latex' );
set( 0, 'defaultLegendInterpreter',        'latex' );
set( 0, 'defaultTextInterpreter',          'latex' );
set( 0, 'defaultAxesFontSize', 12 );
format long 

% Newton parameters
x0 = 2;
nPas = 6;
global approx approxType h;
h = 1e-1;
vPas = 1:(nPas+1);

% Using exact derivative
approx = false;
approxType = 'NONE';
xk = zeros(1, nPas+1);
xk(1) = x0;
for k = 1:nPas
    xk(k+1) = xk(k) - f(xk(k))/fPrim(xk(k));
end
semilogy(vPas, abs(xk-pi), '-o', 'DisplayName', "f' exact");
hold on;

% Using approximate constant derivative
approx = true;
approxType = 'EVAL';
xk = zeros(1, nPas+1);
xk(1) = x0;
for k = 1:nPas
    xk(k+1) = xk(k) - f(xk(k))/fPrim(xk(k));
end
semilogy(vPas, abs(xk-pi), '-s', 'DisplayName', "f' const");

% Using finite difference approximation
approxType = 'FINDIFF';
xk = zeros(1, nPas+1);
xk(1) = x0;
for k = 1:nPas
    xk(k+1) = xk(k) - f(xk(k))/fPrim(xk(k));
end
semilogy(vPas, abs(xk-pi), '-^', 'DisplayName', "f' finDiff");

grid on; xlabel('$N_{iter}$'); ylabel('error'); legend()


function sol = f(x)
    sol = sin(x);
end

function sol = fPrim(x)
    global approx approxType h; 
    if approx
        if strcmp(approxType,'FINDIFF')
            sol = (f(x+h)-f(x-h))/(2*h);
        elseif strcmp(approxType, 'EVAL')
            sol = -0.90;
        end
    else
        sol = cos(x);
    end
end

File: chap1/sec1.5/prob1.10/sol_findPi.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Fri Sep 28 16:16:06 2018

@author: lunet
"""
import numpy as np
import matplotlib.pyplot as plt

# Main parameters
x0 = 2
nPas = 6
h = 1e-1
approx = True
approxType = 'FINDIFF'


# Function and derivative
def f(x):
    return np.sin(x)


def fPrim(x):
    if approx:
        if approxType == 'FINDIFF':
            return (f(x+h)-f(x-h))/(2*h)
        elif approxType == 'EVAL':
            return -0.9
    else:
        return np.cos(x)


# Newton with exact derivative
approx = False
xk = np.zeros(nPas+1)
xk[0] = x0
for k in range(nPas):
    xk[k+1] = xk[k] - f(xk[k])/fPrim(xk[k])
plt.semilogy(np.arange(nPas+1), abs(xk-np.pi), 'o-', label="f' exact")

# Newton with approximated constant derivative
approx = True
approxType = 'EVAL'
xk = np.zeros(nPas+1)
xk[0] = x0
for k in range(nPas):
    xk[k+1] = xk[k] - f(xk[k])/fPrim(xk[k])
plt.semilogy(np.arange(nPas+1), abs(xk-np.pi), 's-', label="f' const")

# Newton with finite difference derivative
approxType = 'FINDIFF'
xk = np.zeros(nPas+1)
xk[0] = x0
for k in range(nPas):
    xk[k+1] = xk[k] - f(xk[k])/fPrim(xk[k])
plt.semilogy(np.arange(nPas+1), abs(xk-np.pi), '^-', label="f' finDiff")

# Plotting parameters
plt.grid()
plt.xlabel('$N_{iter}$')
plt.ylabel('error')
plt.legend()
plt.savefig('newtonPi.pdf', bbox_inches='tight')
plt.tight_layout()
plt.show()
